File: cef-java-32/src/main/java/com/ui/cef_control/lifecycle/ProcessController.java
package com.anca.appl.fw.gui.cef_control.lifecycle;

import java.util.ArrayList;
import java.util.List;

public class ProcessController {

	public enum State {
		NEW,        // never started
		STARTING,   // start requested
		RUNNING,    // fully started
		STOPPING,  // controlled shutdown
		STOPPED,   // cleanly stopped
		ERROR       // crashed
	}

	private State state = State.NEW;
	private final List<LifecycleListener> listeners;

	public ProcessController() {
		this.state = State.NEW;
		this.listeners = new ArrayList<>();
	}

	public void addListener(LifecycleListener listener) {
		if (listener != null && !listeners.contains(listener)) {
			listeners.add(listener);
		}
	}

	public void removeListener(LifecycleListener listener) {
		listeners.remove(listener);
	}

	public void start() {
		if (state != State.NEW && state != State.STOPPED) {
			throw new IllegalStateException(
					"Cannot start in state " + state
			);
		}

		state = State.STARTING;
		notifyStarted();
		state = State.RUNNING;
	}

	public void stop() {
		if (state != State.RUNNING) {
			throw new IllegalStateException(
					"Cannot stop in state " + state
			);
		}

		state = State.STOPPING;
		notifyStopping();

		state = State.STOPPED;
		notifyStopped();
	}

	public void onError(Throwable error) {
		if (state == State.ERROR || state == State.STOPPED) {
			return; // idempotent
		}

		state = State.ERROR;
		notifyError(error);

		state = State.STOPPED;
		notifyStopped();
	}

	public void restart() {
		if (state != State.STOPPED && state != State.ERROR) {
			throw new IllegalStateException(
					"Cannot restart in state " + state
			);
		}

		state = State.STARTING;
		notifyStarted();
		state = State.RUNNING;
	}

	public State getState() {
		return state;
	}

	private void notifyStarted() {
		for (LifecycleListener listener : listeners) {
			listener.onStarted();
		}
	}

	private void notifyStopping() {
		for (LifecycleListener listener : listeners) {
			listener.onStopping();
		}
	}

	private void notifyStopped() {
		for (LifecycleListener listener : listeners) {
			listener.onStopped();
		}
	}

	private void notifyError(Throwable error) {
		for (LifecycleListener listener : listeners) {
			listener.onError(error);
		}
	}
}

File: cef-parallel/src/core/ProcessController.cpp
#include "../../inc/core/ProcessController.h"
#include <algorithm>

namespace cef_ui {
namespace core {

ProcessController::ProcessController()
    : state_(ProcessState::Idle),
      listeners_() {}

void ProcessController::Start() {
  if (state_ != ProcessState::Idle) {
    throw std::logic_error("ProcessController::Start() called but process is not Idle");
  }

  state_ = ProcessState::Starting;
  state_ = ProcessState::Started;
  EmitStarted();
}

void ProcessController::Shutdown() {
  // Idempotent: shutdown is always safe, even if already stopped
  if (state_ == ProcessState::Stopped) {
    return;
  }

  // Only emit events if we're actually shutting down from a running state
  if (state_ == ProcessState::Started) {
    state_ = ProcessState::Stopping;
    EmitStopping();
    state_ = ProcessState::Stopped;
    EmitStopped();
  } else {
    // If in Idle or other states, just transition to Stopped without events
    state_ = ProcessState::Stopped;
  }
}

ProcessState ProcessController::GetState() const {
  return state_;
}

void ProcessController::AddListener(std::shared_ptr<ILifecycleListener> listener) {
  if (listener) {
    listeners_.push_back(listener);
  }
}

void ProcessController::RemoveListener(std::shared_ptr<ILifecycleListener> listener) {
  listeners_.erase(
      std::remove(listeners_.begin(), listeners_.end(), listener),
      listeners_.end());
}

void ProcessController::EmitStarted() {
  for (auto& listener : listeners_) {
    listener->OnStarted();
  }
}

void ProcessController::EmitStopping() {
  for (auto& listener : listeners_) {
    listener->OnStopping();
  }
}

void ProcessController::EmitStopped() {
  for (auto& listener : listeners_) {
    listener->OnStopped();
  }
}

void ProcessController::EmitError(const std::string& error_message) {
  for (auto& listener : listeners_) {
    listener->OnError(error_message);
  }
}

}  // namespace core
}  // namespace cef_ui
